File: frontend/src/types/transaction.ts
/**
 * TypeScript types for transactions, receipts, and categories
 */

export enum PaymentMethod {
  HSA_CARD = 'hsa_card',
  PERSONAL = 'personal',
  OTHER = 'other',
}

export enum ReimbursementStatus {
  NOT_NEEDED = 'not_needed',
  PENDING = 'pending',
  REIMBURSED = 'reimbursed',
  NOT_SEEKING = 'not_seeking',
}

export interface Transaction {
  id: string
  family_id: string
  family_member_id: string
  category_id: string
  hsa_account_id?: string
  transaction_date: string // ISO date string
  amount: number
  merchant_name?: string
  description?: string
  payment_method: PaymentMethod
  reimbursement_status: ReimbursementStatus
  reimbursed_date?: string // ISO date string
  tax_year: number
  created_by_user_id?: string
  created_at: string
  updated_at: string
}

export interface TransactionCreate {
  family_member_id: string
  category_id: string
  transaction_date: string // ISO date string
  amount: number
  merchant_name?: string
  description?: string
  payment_method: PaymentMethod
  reimbursement_status: ReimbursementStatus
  reimbursed_date?: string // ISO date string
  tax_year?: number
  hsa_account_id?: string
}

export interface TransactionUpdate {
  family_member_id?: string
  category_id?: string
  transaction_date?: string
  amount?: number
  merchant_name?: string
  description?: string
  payment_method?: PaymentMethod
  reimbursement_status?: ReimbursementStatus
  reimbursed_date?: string
  tax_year?: number
  hsa_account_id?: string
}

export interface TransactionListResponse {
  transactions: Transaction[]
  total: number
  skip: number
  limit: number
}

export interface TransactionFilters {
  family_member_id?: string
  category_id?: string
  start_date?: string
  end_date?: string
  min_amount?: number
  max_amount?: number
  payment_method?: PaymentMethod
  reimbursement_status?: ReimbursementStatus
  tax_year?: number
  search?: string
  skip?: number
  limit?: number
}

export interface Category {
  id: string
  family_id?: string
  name: string
  description?: string
  is_hsa_eligible: boolean
  parent_category_id?: string
  is_system_category: boolean
  created_at: string
}

export interface CategoryCreate {
  name: string
  description?: string
  is_hsa_eligible: boolean
  parent_category_id?: string
}

export interface Receipt {
  id: string
  transaction_id: string
  file_name: string
  file_size: number
  mime_type: string
  s3_key: string
  s3_bucket: string
  thumbnail_s3_key?: string
  upload_date: string
  uploaded_by_user_id?: string
  created_at: string
}

export interface ReceiptUploadUrl {
  upload_url: string
  s3_key: string
  fields: Record<string, string>
  transaction_id: string
}

export interface ReceiptCreate {
  file_name: string
  file_size: number
  mime_type: string
  s3_key: string
  s3_bucket: string
  thumbnail_s3_key?: string
}

export interface FamilyMember {
  id: string
  family_id: string
  user_id?: string
  name: string
  relationship: 'self' | 'spouse' | 'child' | 'other'
  date_of_birth?: string
  is_tax_dependent: boolean
  is_active: boolean
  created_at: string
  updated_at: string
}

// Helper functions for display
export const getPaymentMethodLabel = (method: PaymentMethod): string => {
  switch (method) {
    case PaymentMethod.HSA_CARD:
      return 'HSA Card'
    case PaymentMethod.PERSONAL:
      return 'Personal Card/Cash'
    case PaymentMethod.OTHER:
      return 'Other'
    default:
      return method
  }
}

export const getReimbursementStatusLabel = (status: ReimbursementStatus): string => {
  switch (status) {
    case ReimbursementStatus.NOT_NEEDED:
      return 'Not Needed'
    case ReimbursementStatus.PENDING:
      return 'Pending Reimbursement'
    case ReimbursementStatus.REIMBURSED:
      return 'Reimbursed'
    case ReimbursementStatus.NOT_SEEKING:
      return 'Not Seeking Reimbursement'
    default:
      return status
  }
}

export const getReimbursementStatusColor = (status: ReimbursementStatus): string => {
  switch (status) {
    case ReimbursementStatus.NOT_NEEDED:
      return 'text-gray-600 bg-gray-100'
    case ReimbursementStatus.PENDING:
      return 'text-yellow-700 bg-yellow-100'
    case ReimbursementStatus.REIMBURSED:
      return 'text-green-700 bg-green-100'
    case ReimbursementStatus.NOT_SEEKING:
      return 'text-blue-700 bg-blue-100'
    default:
      return 'text-gray-600 bg-gray-100'
  }
}

File: backend/app/schemas/transaction.py
"""Transaction and Receipt Pydantic schemas"""

from datetime import date, datetime
from decimal import Decimal
from typing import Optional
from uuid import UUID
from pydantic import BaseModel, Field, validator

from app.models.transaction import PaymentMethod, ReimbursementStatus


# Transaction Schemas

class TransactionBase(BaseModel):
    """Base transaction schema"""
    family_member_id: UUID
    category_id: UUID
    transaction_date: date
    amount: Decimal = Field(gt=0, decimal_places=2)
    merchant_name: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = None
    payment_method: PaymentMethod = PaymentMethod.PERSONAL
    reimbursement_status: ReimbursementStatus = ReimbursementStatus.PENDING
    reimbursed_date: Optional[date] = None
    tax_year: Optional[int] = None
    hsa_account_id: Optional[UUID] = None

    @validator("tax_year", pre=True, always=True)
    def set_tax_year(cls, v, values):
        """Auto-set tax year from transaction date if not provided"""
        if v is None and "transaction_date" in values:
            return values["transaction_date"].year
        return v

    @validator("reimbursed_date")
    def validate_reimbursed_date(cls, v, values):
        """Ensure reimbursed_date is set when status is reimbursed"""
        if values.get("reimbursement_status") == ReimbursementStatus.REIMBURSED and v is None:
            raise ValueError("reimbursed_date is required when status is reimbursed")
        return v


class TransactionCreate(TransactionBase):
    """Schema for creating a transaction"""
    pass


class TransactionUpdate(BaseModel):
    """Schema for updating a transaction (all fields optional)"""
    family_member_id: Optional[UUID] = None
    category_id: Optional[UUID] = None
    transaction_date: Optional[date] = None
    amount: Optional[Decimal] = Field(None, gt=0, decimal_places=2)
    merchant_name: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = None
    payment_method: Optional[PaymentMethod] = None
    reimbursement_status: Optional[ReimbursementStatus] = None
    reimbursed_date: Optional[date] = None
    tax_year: Optional[int] = None
    hsa_account_id: Optional[UUID] = None


class TransactionResponse(TransactionBase):
    """Schema for transaction response"""
    id: UUID
    family_id: UUID
    created_by_user_id: Optional[UUID]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# Receipt Schemas

class ReceiptBase(BaseModel):
    """Base receipt schema"""
    file_name: str = Field(..., max_length=255)
    file_size: int = Field(..., gt=0)
    mime_type: str = Field(..., max_length=100)


class ReceiptCreate(ReceiptBase):
    """Schema for creating a receipt"""
    s3_key: str = Field(..., max_length=512)
    s3_bucket: str = Field(..., max_length=255)
    thumbnail_s3_key: Optional[str] = Field(None, max_length=512)


class ReceiptResponse(ReceiptBase):
    """Schema for receipt response"""
    id: UUID
    transaction_id: UUID
    s3_key: str
    s3_bucket: str
    thumbnail_s3_key: Optional[str]
    upload_date: datetime
    uploaded_by_user_id: Optional[UUID]
    created_at: datetime

    class Config:
        from_attributes = True


# Category Schemas

class CategoryBase(BaseModel):
    """Base category schema"""
    name: str = Field(..., max_length=255)
    description: Optional[str] = None
    is_hsa_eligible: bool = True
    parent_category_id: Optional[UUID] = None


class CategoryCreate(CategoryBase):
    """Schema for creating a custom category"""
    pass


class CategoryResponse(CategoryBase):
    """Schema for category response"""
    id: UUID
    family_id: Optional[UUID]  # None for system categories
    is_system_category: bool
    created_at: datetime

    class Config:
        from_attributes = True


# Query/Filter Schemas

class TransactionFilters(BaseModel):
    """Filters for transaction list"""
    family_member_id: Optional[UUID] = None
    category_id: Optional[UUID] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    min_amount: Optional[Decimal] = None
    max_amount: Optional[Decimal] = None
    payment_method: Optional[PaymentMethod] = None
    reimbursement_status: Optional[ReimbursementStatus] = None
    tax_year: Optional[int] = None
    search: Optional[str] = None  # Search merchant or description
    skip: int = Field(0, ge=0)
    limit: int = Field(100, ge=1, le=1000)


class TransactionListResponse(BaseModel):
    """Response for transaction list with pagination"""
    transactions: list[TransactionResponse]
    total: int
    skip: int
    limit: int
